MudAutocomplete, MudSelect: Fix Focus and Clear

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Resolves #10733 by allowing the Clear button to actually clear the values
Resolves #10705 by reordering the focus logic
Resolves #10718 by removing additional click handler
Resolves #10739

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Unit tests passing after changing Click(s) to MouseDown(s), Visual tests done with video in comments.

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
Revert "Build: Fix for missing MudBlazor.min.css in SassCompiler (#10643)"

This reverts commit 10a60e0b02dbc9e43391455d36e19aeea76fb1f8.

versile2 committed:
viewer example

versile2 committed:
Merge branch 'MudBlazor:dev' into fix/autocompletefocus

versile2 committed:
Merge branch 'MudBlazor:dev' into fix/autocompletefocus

versile2 committed:
Re-establish opening order for selecting an item

versile2 committed:
add handler for clear button click

versile2 committed:
mimic SelectClear function

Anu6is commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -525,8 +525,14 @@ public async Task SelectOptionAsync(T value)
                 await SetValueAsync(value);
 
+                // needs to be open to run the rest of the code
+                Open = true;
```
How does this open/close look in the UI?

versile2 replied:
// Note: setting Open will not open or close it. Use ToggleMenuAsync() for that.

It has no visual changes

Anu6is commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -973,6 +980,19 @@ private async Task OnInputActivationAsync(bool openMenu)
+        {
+            // clear button clicked, let's make sure text is cleared
+            await ClearAsync();
+            await BeginValidateAsync();
```
Why the call to validate after clearing?

versile2 replied:
1) To match the way MudSelect works
2) If it's being a validated field or required, this will show an error as any other field would once they are touched.

danielchalmers commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor:
```diff
@@ -39,7 +39,7 @@
                           AdornmentAriaLabel="@AdornmentAriaLabel"
                           Clearable="@(Clearable && !GetReadOnlyState())"
-                          OnClearButtonClick="@OnClearButtonClick"
+                          OnClearButtonClick="(async (e) => await HandleClearButtonClicked(e))"
```
```suggestion
                          OnClearButtonClick="HandleClearButtonAsync"
```

versile2 replied:
I matched MudSelect, prefer they all match?
```
                          ClearIcon="@ClearIcon"
                          OnClearButtonClick="@(async (e) => await SelectClearButtonClickHandlerAsync(e))"
                          @attributes="UserAttributes"
```

danielchalmers commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -973,6 +980,19 @@ private async Task OnInputActivationAsync(bool openMenu)
             }
         }
 
+        internal async ValueTask HandleClearButtonClicked(MouseEventArgs e)
```
```suggestion
        internal async Task HandleClearButtonAsync(MouseEventArgs e)
```

danielchalmers reviewed (changes requested):
small api tweak to match mudmask

danielchalmers commented:
Are there any other controls that have a clear button that are afflicted by this?

versile2 committed:
Update src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs

from ValueTask to Task

Co-authored-by: Daniel Chalmers <daniel.chalmers@outlook.com>

versile2 committed:
aligned method name

danielchalmers commented:
Sorry, is AutocompleteRetainFocusTest the right example or is it for the other PR? I'm seeing a message box when I try to clear the field.

https://github.com/user-attachments/assets/6308b8d8-0be6-4d38-bf2d-fc8ab1ee7920

danielchalmers commented:
@ScarletKuro Could you take a look at the property logic when you get a chance?

danielchalmers commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor:
```diff
@@ -39,7 +39,7 @@
                           AdornmentAriaLabel="@AdornmentAriaLabel"
                           Clearable="@(Clearable && !GetReadOnlyState())"
-                          OnClearButtonClick="@OnClearButtonClick"
+                          OnClearButtonClick="(async (e) => await HandleClearButtonAsync(e))"
```
```suggestion
                          OnClearButtonClick="HandleClearButtonAsync"
```

versile2 commented:
> Sorry, is AutocompleteRetainFocusTest the right example or is it for the other PR? I'm seeing a message box when I try to clear the field.
> Video3.mp4

No just the Docs autocomplete first example is what I used here. I did go ahead and add the logic in autocompleterefocustest to prevent popup when the value is empty

versile2 committed:
fix viewer test and change the input for handle

danielchalmers reviewed (approved):
Working on my end 👍

versile2 committed:
Merge branch 'MudBlazor:dev' into fix/autocompletefocus

versile2 committed:
changed focus area, did not reuse ClearAsync

versile2 committed:
fix bug from #10718 -onmousedown + onclick some browsers fire twice

versile2 committed:
updated example

versile2 committed:
repositioned focus

versile2 committed:
Better Test Example

versile2 committed:
Unit Test adjustments from Click to MouseDown

versile2 commented:
Added fixes for #10718 as well as tested on all browsers. Feel free to re-review.


https://github.com/user-attachments/assets/4c452952-6f81-498c-9073-a7724a4c7f6e

danielchalmers reviewed (approved):
Good call on the click/mousedown events. I wasn't aware that the browser synthesizes the mousedown event on a full tap in contrast to pointerdown that works on tap start

ScarletKuro commented on src/MudBlazor.UnitTests.Viewer/TestComponents/Autocomplete/AutocompleteRetainFocusTest.razor:
```diff
@@ -0,0 +1,145 @@
+Clearable="true" />
+<MudText>@($"Text: {autocomp1?.Text} Value: {autocomp1?.Value} Local Value: {_messageBoxValue}")</MudText>
+
+<MudSelect T="string" @ref=select1
```
The formatting on the whole file is incorrect.

versile2 replied:
dunno why my IDE does that sometimes. fixed though.

versile2 committed:
format

ScarletKuro commented:
I hope we've made sure that everything works with both mouse and touch devices for the MudSelect and MudAutocomplete, because I'm starting to get confused. In v8, we changed from `onclick` to `onpointerdown`, then from `onpointerdown` to `onmousedown` and `onclick`, and now from `onmousedown` and `onclick` to just `onmousedown`. If people still complain, let's just revert it to how it was in v7, I guess.

I'm also slightly confused by this new method:
https://github.com/MudBlazor/MudBlazor/blob/c02c66b021b0aa5edb81b4b61e9f8d1a78555101/src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs#L983  
Wasn't the `clear` functionality working before without any additional methods? I thought it was just a problem with `stopPropagation`, like it was here https://github.com/MudBlazor/MudBlazor/pull/10737 and here https://github.com/MudBlazor/MudBlazor/pull/10353  
I'm just worried this might mess up the validation.

versile2 commented:
> I hope we've made sure that everything works with both mouse and touch devices for the MudSelect and MudAutocomplete, because I'm starting to get confused. In v8, we changed from `onclick` to `onpointerdown`, then from `onpointerdown` to `onmousedown` and `onclick`, and now from `onmousedown` and `onclick` to just `onmousedown`. If people still complain, let's just revert it to how it was in v7, I guess.
> 
> I'm also slightly confused by this new method:
> 
> https://github.com/MudBlazor/MudBlazor/blob/c02c66b021b0aa5edb81b4b61e9f8d1a78555101/src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs#L983
> 
> Wasn't the Clear functionality working before without any additional methods? I thought it was just a problem with `stopPropagation`, like it was here #10737 and here #10353
> I'm just worried this might mess up the validation.

I don't think so tbh, the reason I added Value/Text/bound Value underneath each is because ClearAsync method doesn't actually clear. And it has commented workarounds in it so I just copied the functionality like Select does it and left that method alone. I honestly don't use autocomplete in my projects so I'm unsure if clear actually worked right. Without the added method hitting the clear button only cleared the text so you could type something else and search. It left the Value in so if you didn't type something new it just reverts because it never fires the ValueChanged event.

danielchalmers commented:
> I hope we've made sure that everything works with both mouse and touch devices for the MudSelect and MudAutocomplete, because I'm starting to get confused. In v8, we changed from `onclick` to `onpointerdown`, then from `onpointerdown` to `onmousedown` and `onclick`, and now from `onmousedown` and `onclick` to just `onmousedown`. If people still complain, let's just revert it to how it was in v7, I guess.

`onpointerdown` triggers when you start touching it

`onmousedown` was included with `onclick` with the assumption that the former would trigger on mouse and the latter otherwise, but firefox handles it differently. `onmousedown` alone works in this case because it is triggered as part of a full tap as well as the mouse click starting

ScarletKuro commented on src/MudBlazor.UnitTests.Viewer/TestComponents/Autocomplete/AutocompleteRetainFocusTest.razor:
```diff
@@ -0,0 +1,145 @@
+        </DialogActions>
+    </MudDialog>
+
+    @code {
```
Still formatting looks wrong, the `@code` and the `MudSelect` tag ends while the others after have an weird indent

versile2 replied:
I ended up having to do it manually. Frustrating when VS does that

versile2 committed:
formatted manually

ScarletKuro reviewed (approved):
LGTM, I haven't tested personally so feel free to merge if ya confident
